polkadot/runtime/parachains/src/mock.rs

```diff
@@ -388,6 +389,142 @@ impl ValidatorSetWithIdentification<AccountId> for MockValidatorSet {
 	type IdentificationOf = FoolIdentificationOf;
 }
 
+pub mod mock_assigner {
+	use super::*;
+	use assigner::UnifiedAssignment;
+	pub use pallet::*;
+
+	pub type TestUnifiedAssignment = UnifiedAssignment<V0Assignment, V0Assignment>;
+
+	#[frame_support::pallet]
+	pub mod pallet {
+		use super::*;
+
+		#[pallet::pallet]
+		#[pallet::without_storage_info]
+		pub struct Pallet<T>(_);
+
+		#[pallet::config]
+		pub trait Config: frame_system::Config + configuration::Config + paras::Config {}
+
+		#[pallet::storage]
+		pub(super) type MockAssignerQueue<T: Config> =
+			StorageValue<_, VecDeque<V0Assignment>, ValueQuery>;
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn add_on_demand_order(
+			order: TestUnifiedAssignment,
+			location: QueuePushDirection,
+		) -> Result<(), DispatchError> {
+			match order {
+				TestUnifiedAssignment::OnDemand(inner) => {
+					MockAssignerQueue::<T>::try_mutate(|queue| {
+						// Don't need to worry about max assignment queue size in scheduler tests
+						match location {
+							QueuePushDirection::Back => queue.push_back(inner),
+							QueuePushDirection::Front => queue.push_front(inner),
+						};
+						Ok(())
+					})
+				}
+				_ => Ok(()),
+			}
+		}
+
+		// Wraps queued assignments as TestUnifiedAssignments so we only have to deal with one
+		// assignment type in tests
+		pub fn get_queue() -> VecDeque<TestUnifiedAssignment> {
+			MockAssignerQueue::<T>::get()
+				.clone()
+				.into_iter()
+				.map(|assignment| TestUnifiedAssignment::OnDemand(assignment))
+				.collect()
+		}
+
+		fn is_legacy_core(core_idx: &CoreIndex) -> bool {
+			let parachain_cores = <paras::Pallet<Test>>::parachains().len() as u32;
+			(0..parachain_cores).contains(&core_idx.0)
+		}
+	}
+
+	impl<T: Config> AssignmentProvider<BlockNumber> for Pallet<T> {
+		type AssignmentType = TestUnifiedAssignment;
```

Comment:
The main reason is so that `push_back_assignment` knows how to behave. It is a part of the `AssignmentProvider` interface, so I can't change its signature. And it only has the parameter `AssignmentType` to tell it how to behave. Using an `AssignmentType` enum seems elegant enough though perhaps it could be made simpler still. A bunch of tests were failing before I swapped in `TestUnifiedAssignment`.

I'm open to other options if there's something you'd prefer!

Reply:
It occurs to me that checking whether the ParaId in a V0Assignment corresponds to a parachain would likely be sufficient. I'll try that, and get rid of `TestUnifiedAssignment` if so.

Reply:
The point is: This interface does not tell anything about the nature of the assignment provider. If assignments would need to be of a certain type, the architecture would be completely broken. The only requirement is that it implements `Assignment`.